File: csharp/allergies/Allergies.cs
using System;
using System.Collections.Generic;

[Flags]
public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128
}

public class Allergies
{
    private Allergen allergies;

    public Allergies(int mask)
    {
        allergies = (Allergen)unchecked((byte)mask);
    }

    public bool IsAllergicTo(Allergen allergen) => allergies.HasFlag(allergen);

    public Allergen[] List()
    {
        var result = new List<Allergen>();

        if (allergies.HasFlag(Allergen.Eggs))
        {
            result.Add(Allergen.Eggs);
        }

        if (allergies.HasFlag(Allergen.Peanuts))
        {
            result.Add(Allergen.Peanuts);
        }

        if (allergies.HasFlag(Allergen.Shellfish))
        {
            result.Add(Allergen.Shellfish);
        }

        if (allergies.HasFlag(Allergen.Strawberries))
        {
            result.Add(Allergen.Strawberries);
        }

        if (allergies.HasFlag(Allergen.Tomatoes))
        {
            result.Add(Allergen.Tomatoes);
        }

        if (allergies.HasFlag(Allergen.Chocolate))
        {
            result.Add(Allergen.Chocolate);
        }

        if (allergies.HasFlag(Allergen.Pollen))
        {
            result.Add(Allergen.Pollen);
        }

        if (allergies.HasFlag(Allergen.Cats))
        {
            result.Add(Allergen.Cats);
        }

        return result.ToArray();
    }
}

File: typescript/allergies/allergies.ts
enum AllergyList {
  none          = 0,      // 0
  eggs          = 1 << 0, // 1
  peanuts       = 1 << 1, // 2
  shellfish     = 1 << 2, // 4
  strawberries  = 1 << 3, // 8
  tomatoes      = 1 << 4, // 16
  chocolate     = 1 << 5, // 32
  pollen        = 1 << 6, // 64
  cats          = 1 << 7, // 128
}

type Allergy = keyof typeof AllergyList;

export class Allergies {
  private readonly allergies: AllergyList

  constructor(allergenIndex: number) {
    this.allergies = <AllergyList>allergenIndex
  }

  public list(): string[] {
    const bits = this.allergyBits
    return bits
      .map(bit => this.has(bit)? AllergyList[bit] : '')
      .filter(text => text);
  }

  private has(allergy: AllergyList): boolean {
    return (this.allergies & allergy) === allergy;
  }

  private get allergyBits(): number[] {
    return Object.keys(AllergyList).map(Number).filter(Boolean);
  }

  public allergicTo(allergen: Allergy): boolean {
    return this.has(AllergyList[allergen]);
  }
}
